cloud-init.1
.TH CLOUD-INIT 1

.SH NAME
cloud-init \- Cloud instance initialization

.SH SYNOPSIS
.BR "cloud-init" " [-h] [-d] [-f FILES] [--force] [-v] {init,modules,single,query,dhclient-hook,features,analyze,collect-logs,clean,status}"

.SH DESCRIPTION
Cloud-init provides a mechanism for cloud instance initialization.
This is done by identifying the cloud platform that is in use, reading
provided cloud metadata and optional vendor and user
data, and then intializing the instance as requested.

Generally, this command is not normally meant to be run directly by
the user. However, some subcommands may useful to development or
debug of deployments.

.SH OPTIONS
.TP
.B "-h, --help"
Show help message and exit

.TP
.B "-d, --debug"
Show additional pre-action logging (default: False)

.TP
.B "-f <files>, --files <files>"
Additional YAML configuration files to use

.TP
.B "--force"
Force running even if no datasource is found (use at your own risk)

.TP
.B "-v, --version"
Show program's version number and exit

.SH SUBCOMMANDS
Please see the help output for each subcommand for additional details,
flags, and subcommands.

.TP
.B "analyze"
Analyze cloud-init logs and data.

.TP
.B "collect-logs"
Collect and tar all cloud-init debug info.

.TP
.B "clean"
Remove logs and artifacts so cloud-init can re-run.

.TP
.B "dhclient-hook"
Run the dhclient hook to record network info.

.TP
.B "features"
List defined features.

.TP
.B "init"
Initializes cloud-init and performs initial modules.

.TP
.B "modules"
Activates modules using a given configuration key.

.TP
.B "query"
Query standardized instance metadata from the command line.

.TP
.B "single"
Run a single module.

.TP
.B "status"
Report cloud-init status or wait on completion.

.SH COPYRIGHT
Copyright (C) 2020 Canonical Ltd.


ChangeLog
19.3
 - azure: support matching dhcp route-metrics for dual-stack ipv4 ipv6
   (LP: #1850308)
 - configdrive: fix subplatform config-drive for /config-drive source
   [David Kindred] (LP: #1849731)
 - DataSourceSmartOS: reconfigure network on each boot
   [Mike Gerdts] (LP: #1765801)
 - Add config for ssh-key import and consuming user-data [Pavel Zakharov]
 - net: fix subnet_is_ipv6() for stateless|stateful
   [Harald Jensås] (LP: #1848690)
 - OVF: disable custom script execution by default [Xiaofeng Wang]
 - cc_puppet: Implement csr_attributes.yaml support [Matthias Baur]
 - cloud-init.service: on centos/fedora/redhat wait on NetworkManager.service
   (LP: #1843334)
 - azure: Do not lock user on instance id change [Sam Eiderman] (LP: #1849677)
 - net/netplan: use ipv6-mtu key for specifying ipv6 mtu values
 - Fix usages of yaml, and move yaml_dump to safeyaml.dumps. (LP: #1849640)
 - exoscale: Increase url_max_wait to 120s. [Chris Glass]
 - net/sysconfig: fix available check on SUSE distros
   [Robert Schweikert] (LP: #1849378)
 - docs: Fix incorrect Azure IMDS IP address [Joshua Powers] (LP: #1849508)
 - introduce .travis.yml
 - net: enable infiniband support in eni and sysconfig renderers
   [Darren Birkett] (LP: #1847114)
 - guestcust_util: handle special characters in config file [Xiaofeng Wang]
 - fix some more typos in comments [Dominic Schlegel]
 - replace any deprecated log.warn with log.warning
   [Dominic Schlegel] (LP: #1508442)
 - net: handle openstack dhcpv6-stateless configuration
   [Harald Jensås] (LP: #1847517)
 - Add .venv/ to .gitignore [Dominic Schlegel]
 - Small typo fixes in code comments. [Dominic Schlegel]
 - cloud_test/lxd: Retry container delete a few times
 - Add Support for e24cloud to Ec2 datasource. (LP: #1696476)
 - Add RbxCloud datasource [Adam Dobrawy]
 - get_interfaces: don't exclude bridge and bond members (LP: #1846535)
 - Add support for Arch Linux in render-cloudcfg [Conrad Hoffmann]
 - util: json.dumps on python 2.7 will handle UnicodeDecodeError on binary
   (LP: #1801364)
 - debian/ubuntu: add missing word to netplan/ENI header (LP: #1845669)
 - ovf: do not generate random instance-id for IMC customization path
 - sysconfig: only write resolv.conf if network_state has DNS values
   (LP: #1843634)
 - sysconfig: use distro variant to check if available (LP: #1843584)
 - systemd/cloud-init.service.tmpl: start after wicked.service
   [Robert Schweikert]
 - docs: fix zstack documentation lints
 - analyze/show: remove trailing space in output
 - Add missing space in warning: "not avalid seed" [Brian Candler]
 - pylintrc: add 'enter_context' to generated-members list
 - Add datasource for ZStack platform. [Shixin Ruan] (LP: #1841181)
 - docs: organize TOC and update summary of project [Joshua Powers]
 - tools: make clean now cleans the dev directory, not the system
 - docs: create cli specific page [Joshua Powers]
 - docs: added output examples to analyze.rst [Joshua Powers]
 - docs: doc8 fixes for instancedata page [Joshua Powers]
 - docs: clean up formatting, organize boot page [Joshua Powers]
 - net: add is_master check for filtering device list (LP: #1844191)
 - docs: more complete list of availability [Joshua Powers]
 - docs: start FAQ page [Joshua Powers]
 - docs: cleanup output & order of datasource page [Joshua Powers]
 - Brightbox: restrict detection to require full domain match .brightbox.com
 - VMWware: add option into VMTools config to enable/disable custom script.
   [Xiaofeng Wang]
 - net,Oracle: Add support for netfailover detection
 - atomic_helper: add DEBUG logging to write_file (LP: #1843276)
 - doc: document doc, create makefile and tox target [Joshua Powers]
 - .gitignore: ignore files produced by package builds
 - docs: fix whitespace, spelling, and line length [Joshua Powers]
 - docs: remove unnecessary file in doc directory [Joshua Powers]
 - Oracle: Render secondary vnic IP and MTU values only
 - exoscale: fix sysconfig cloud_config_modules overrides (LP: #1841454)
 - net/cmdline: refactor to allow multiple initramfs network config sources
 - ubuntu-drivers: call db_x_loadtemplatefile to accept NVIDIA EULA
   (LP: #1840080)
 - Add missing #cloud-config comment on first example in documentation.
   [Florian Müller]
 - ubuntu-drivers: emit latelink=true debconf to accept nvidia eula
   (LP: #1840080)
 - DataSourceOracle: prefer DS network config over initramfs
 - format.rst: add text/jinja2 to list of content types (+ cleanups)
 - Add GitHub pull request template to point people at hacking doc
 - cloudinit/distros/parsers/sys_conf: add docstring to SysConf
 - pyflakes: remove unused variable [Joshua Powers]
 - Azure: Record boot timestamps, system information, and diagnostic events
   [Anh Vo]
 - DataSourceOracle: configure secondary NICs on Virtual Machines
 - distros: fix confusing variable names
 - azure/net: generate_fallback_nic emits network v2 config instead of v1
 - Add support for publishing host keys to GCE guest attributes [Rick Wright]
 - New data source for the Exoscale.com cloud platform [Chris Glass]
 - doc: remove intersphinx extension
 - cc_set_passwords: rewrite documentation (LP: #1838794)
 - net/cmdline: split interfaces_by_mac and init network config determination
 - stages: allow data sources to override network config source order
 - cloud_tests: updates and fixes
 - Fix bug rendering MTU on bond or vlan when input was netplan. (LP: #1836949)
 - net: update net sequence, include wait on netdevs, opensuse netrules path
   (LP: #1817368)
19.2:
 - net: add rfc3442 (classless static routes) to EphemeralDHCP
   (LP: #1821102)
 - templates/ntp.conf.debian.tmpl: fix missing newline for pools
   (LP: #1836598)
 - Support netplan renderer in Arch Linux [Conrad Hoffmann]
 - Fix typo in publicly viewable documentation. [David Medberry]
 - Add a cdrom size checker for OVF ds to ds-identify
   [Pengpeng Sun] (LP: #1806701)
 - VMWare: Trigger the post customization script via cc_scripts module.

cloud-init.py
#!/usr/bin/python
# vi: ts=4 expandtab
#
#    Copyright (C) 2009-2010 Canonical Ltd.
#    Copyright (C) 2012 Hewlett-Packard Development Company, L.P.
#
#    Author: Scott Moser <scott.moser@canonical.com>
#    Author: Juerg Haefliger <juerg.haefliger@hp.com>
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License version 3, as
#    published by the Free Software Foundation.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.

import subprocess
import sys

import cloudinit
import cloudinit.util as util
import cloudinit.CloudConfig as CC
import cloudinit.DataSource as ds
import cloudinit.netinfo as netinfo
import time
import traceback
import logging
import errno
import os


def warn(wstr):
    sys.stderr.write("WARN:%s" % wstr)


def main():
    util.close_stdin()

    cmds = ("start", "start-local")
    deps = {"start": (ds.DEP_FILESYSTEM, ds.DEP_NETWORK),
            "start-local": (ds.DEP_FILESYSTEM, )}

    cmd = ""
    if len(sys.argv) > 1:
        cmd = sys.argv[1]

    cfg_path = None
    if len(sys.argv) > 2:
        # this is really for debugging only
        # but you can invoke on development system with ./config/cloud.cfg
        cfg_path = sys.argv[2]

    if not cmd in cmds:
        sys.stderr.write("bad command %s. use one of %s\n" % (cmd, cmds))
        sys.exit(1)

    now = time.strftime("%a, %d %b %Y %H:%M:%S %z", time.gmtime())
    try:
        uptimef = open("/proc/uptime")
        uptime = uptimef.read().split(" ")[0]
        uptimef.close()
    except IOError as e:
        warn("unable to open /proc/uptime\n")
        uptime = "na"

    cmdline_msg = None
    cmdline_exc = None
    if cmd == "start":
        target = "%s.d/%s" % (cloudinit.system_config,
            "91_kernel_cmdline_url.cfg")
        if os.path.exists(target):
            cmdline_msg = "cmdline: %s existed" % target
        else:
            cmdline = util.get_cmdline()
            try:
                (key, url, content) = cloudinit.get_cmdline_url(
                    cmdline=cmdline)
                if key and content:
                    util.write_file(target, content, mode=0600)
                    cmdline_msg = ("cmdline: wrote %s from %s, %s" %
                        (target, key, url))
                elif key:
                    cmdline_msg = ("cmdline: %s, %s had no cloud-config" %
                        (key, url))
            except Exception:
                cmdline_exc = ("cmdline: '%s' raised exception\n%s" %
                    (cmdline, traceback.format_exc()))
                warn(cmdline_exc)

    try:
        cfg = cloudinit.get_base_cfg(cfg_path)
    except Exception as e:
        warn("Failed to get base config. falling back to builtin: %s\n" % e)
        try:
            cfg = cloudinit.get_builtin_cfg()
        except Exception as e:
            warn("Unable to load builtin config\n")
            raise

    try:
        (outfmt, errfmt) = CC.get_output_cfg(cfg, "init")
        CC.redirect_output(outfmt, errfmt)
    except Exception as e:
        warn("Failed to get and set output config: %s\n" % e)

    cloudinit.logging_set_from_cfg(cfg)
    log = logging.getLogger()

    if cmdline_exc:
        log.debug(cmdline_exc)
    elif cmdline_msg:
        log.debug(cmdline_msg)

    try:
        cloudinit.initfs()
    except Exception as e:
        warn("failed to initfs, likely bad things to come: %s\n" % str(e))

    nonet_path = "%s/%s" % (cloudinit.get_cpath("data"), "no-net")

    if cmd == "start":
        print netinfo.debug_info()

        stop_files = (cloudinit.get_ipath_cur("obj_pkl"), nonet_path)
        # if starting as the network start, there are cases
        # where everything is already done for us, and it makes
        # most sense to exit early and silently
        for f in stop_files:
            try:
                fp = open(f, "r")
                fp.close()
            except:
                continue

            log.debug("no need for cloud-init start to run (%s)\n", f)
            sys.exit(0)
    elif cmd == "start-local":
        # cache is not instance specific, so it has to be purged
        # but we want 'start' to benefit from a cache if
        # a previous start-local populated one
        manclean = util.get_cfg_option_bool(cfg, 'manual_cache_clean', False)
        if manclean:
            log.debug("not purging cache, manual_cache_clean = True")
        cloudinit.purge_cache(not manclean)

        try:
            os.unlink(nonet_path)
        except OSError as e:
            if e.errno != errno.ENOENT:
                raise

    msg = "cloud-init %s running: %s. up %s seconds" % (cmd, now, uptime)
    sys.stderr.write(msg + "\n")
    sys.stderr.flush()

    log.info(msg)

    cloud = cloudinit.CloudInit(ds_deps=deps[cmd])

    try:
        cloud.get_data_source()
    except cloudinit.DataSourceNotFoundException as e:
        sys.stderr.write("no instance data found in %s\n" % cmd)
        sys.exit(0)

    # set this as the current instance
    cloud.set_cur_instance()

    # store the metadata
    cloud.update_cache()

    msg = "found data source: %s" % cloud.datasource
    sys.stderr.write(msg + "\n")
    log.debug(msg)

    # parse the user data (ec2-run-userdata.py)
    try:
        ran = cloud.sem_and_run("consume_userdata", cloudinit.per_instance,
            cloud.consume_userdata, [cloudinit.per_instance], False)
        if not ran:
            cloud.consume_userdata(cloudinit.per_always)
    except:
        warn("consuming user data failed!\n")
        raise

    cfg_path = cloudinit.get_ipath_cur("cloud_config")
    cc = CC.CloudConfig(cfg_path, cloud)

    # if the output config changed, update output and err
    try:
        outfmt_orig = outfmt
        errfmt_orig = errfmt
        (outfmt, errfmt) = CC.get_output_cfg(cc.cfg, "init")
        if outfmt_orig != outfmt or errfmt_orig != errfmt:
            warn("stdout, stderr changing to (%s,%s)" % (outfmt, errfmt))
            CC.redirect_output(outfmt, errfmt)
    except Exception as e:
        warn("Failed to get and set output config: %s\n" % e)

    # send the cloud-config ready event
    cc_path = cloudinit.get_ipath_cur('cloud_config')
    cc_ready = cc.cfg.get("cc_ready_cmd",
        ['initctl', 'emit', 'cloud-config',
         '%s=%s' % (cloudinit.cfg_env_name, cc_path)])
    if cc_ready:
        if isinstance(cc_ready, str):
            cc_ready = ['sh', '-c', cc_ready]
        subprocess.Popen(cc_ready).communicate()

    module_list = CC.read_cc_modules(cc.cfg, "cloud_init_modules")

    failures = []
    if len(module_list):
        failures = CC.run_cc_modules(cc, module_list, log)
    else:
        msg = "no cloud_init_modules to run"
        sys.stderr.write(msg + "\n")
        log.debug(msg)
        sys.exit(0)

    sys.exit(len(failures))

if __name__ == '__main__':
    main()
